rclcpp/test/rclcpp/executors/test_executors.cpp

```diff
@@ -0,0 +1,457 @@
+  ExecutorType executor;
+  executor.add_node(this->node);
+
+  // This needs to be longer than it takes to reach the shutdown call below.
+  // This timeout blocks the test finishing, so it shouldn't be too long.
+  std::future<void> future = std::async(
+    std::launch::async,
+    []() {std::this_thread::sleep_for(20ms);});
+
+  bool spin_exited = false;
+
+  // Long timeout
+  std::thread spinner([&spin_exited, &executor, &future]() {
+      auto ret = executor.spin_until_future_complete(future, 1s);
+      EXPECT_EQ(rclcpp::FutureReturnCode::INTERRUPTED, ret);
+      spin_exited = true;
+    });
+
+  // Do some minimal work
+  this->publisher->publish(std_msgs::msg::Empty());
+  std::this_thread::sleep_for(1ms);
+
+  // Force interruption
+  rclcpp::shutdown();
+
+  // Give it time to exit
+  std::this_thread::sleep_for(10ms);
+  EXPECT_TRUE(spin_exited);
+  spinner.join();
+}
+
+class TestWaitable : public rclcpp::Waitable
+{
+public:
+  TestWaitable()
+  {
+    rcl_guard_condition_options_t guard_condition_options =
+      rcl_guard_condition_get_default_options();
+
+    gc_ = rcl_get_zero_initialized_guard_condition();
+    rcl_ret_t ret = rcl_guard_condition_init(
+      &gc_,
+      rclcpp::contexts::get_global_default_context()->get_rcl_context().get(),
+      guard_condition_options);
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(ret);
+    }
+  }
+
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) override
+  {
+    rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &gc_, NULL);
+    if (RCL_RET_OK != ret) {
+      return false;
+    }
+    ret = rcl_trigger_guard_condition(&gc_);
+    return RCL_RET_OK == ret;
+  }
+
+  bool
+  is_ready(rcl_wait_set_t * wait_set) override
+  {
+    (void)wait_set;
+    return true;
+  }
+
+  void
+  execute() override
+  {
+    count_++;
+    std::this_thread::sleep_for(1ms);
+  }
+
+  size_t
+  get_number_of_ready_guard_conditions() override {return 1;}
+
+  size_t
+  get_count()
+  {
+    return count_;
+  }
+
+private:
+  size_t count_ = 0;
+  rcl_guard_condition_t gc_;
+};
+
+TYPED_TEST(TestExecutorsSpinVariants, spinAll) {
+  using ExecutorType = TypeParam;
+  ExecutorType executor;
+  auto waitable_interfaces = this->node->get_node_waitables_interface();
+  auto my_waitable = std::make_shared<TestWaitable>();
+  waitable_interfaces->add_waitable(my_waitable, nullptr);
+  executor.add_node(this->node);
+
+  // Long timeout, this determines the duration of the test, so not making it too long.
+  // Just long enough for multiple waitables to execute.
+  bool spin_exited = false;
+  std::thread spinner([&spin_exited, &executor, this]() {
+      executor.spin_all(20ms);
+      executor.remove_node(this->node);
+      spin_exited = true;
+    });
+
+  // Do some work for longer the test waitable needs
+  for (int i = 0; i < 100; ++i) {
+    this->publisher->publish(std_msgs::msg::Empty());
+    std::this_thread::sleep_for(1ms);
+    if (spin_exited) {
```

Comment:
Here you could cancel the executor if the condition was met, i.e. `my_waitable->get_count() > 1u`, and therefore you could `spin_all` for a longer period, further reducing the likelihood of a flaky test if the waitable is not called twice within 20ms (not a guarantee 100% of the time on cloud based Windows machines in my experience).

Reply:
Updated this logic, and did the same with spinSome.